src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -64,23 +64,26 @@ public partial class JsonTypeInfo<T>
             CancellationToken cancellationToken,
             object? rootValueBoxed = null)
         {
-            return SerializeAsync(new AsyncSerializationStreamContext(utf8Json, Options), rootValue, cancellationToken, rootValueBoxed);
+            return SerializeAsync(new PooledByteBufferWriter(Options.DefaultBufferSize, utf8Json), rootValue, cancellationToken,
+                (int)(Options.DefaultBufferSize * JsonSerializer.FlushThreshold), rootValueBoxed);
```

Comment:
It seems we could move disposal of the PooledByteBufferWriter to this method now?
```suggestion
            using PooledByteBufferWriter bufferWriter = new(Options.DefaultBufferSize, utf8Json);
            int flushThreshold = (int)(Options.DefaultBufferSize * JsonSerializer.FlushThreshold);
            return SerializeAsync(bufferWriter, rootValue, flushThreshold, cancellationToken, rootValueBoxed);
```

Reply:
Nope, that would result in disposal before the async method completes. And if we `await SerializeAsync` to fix that, then we're likely introducing another state machine.